fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReplacePredicate.java

```diff
@@ -0,0 +1,321 @@
+        ExpressionAnalyzer analyzer = new ReplaceAnalyzer(null, new Scope(ImmutableList.of()), null, false, false);
+        Set<Expression> res = new LinkedHashSet<>();
+        for (T equals : equalSet) {
+            Map<Expression, Expression> replaceMap = new HashMap<>();
+            replaceMap.put(equals, replaceToThis);
+            if (!exprPredicates.containsKey(equals)) {
+                continue;
+            }
+            for (Expression predicate : exprPredicates.get(equals)) {
+                Expression newPredicates = Replacer.INSTANCE.replace(predicate, replaceMap);
+                try {
+                    Expression analyzed = analyzer.analyze(newPredicates);
+                    res.add(analyzed.withInferred(true));
+                } catch (Exception e) {
+                    // has cast error, just not infer and do nothing
+                    continue;
+                }
+            }
+        }
+        return res;
+    }
+
+    /* Extract the equivalence relationship a=b, and when case (d_tinyint as int)=d_int is encountered,
+    remove the cast and extract d_tinyint=d_int
+    EqualPairs is the output parameter and the equivalent pair of predicate derivation input,
+    which is used to ensure that the derivation
+    does not generate repeated equivalent conditions, such as a=b and b=a */
+    private static ImmutableEqualSet<Slot> findEqual(Set<Expression> inputs, Set<Pair<Slot, Slot>> equalPairs) {
+        ImmutableEqualSet.Builder<Slot> fromCastEqualSetBuilder = new ImmutableEqualSet.Builder<>();
+        for (Expression input : inputs) {
+            if (!(input instanceof EqualTo)) {
+                continue;
+            }
+            EqualTo equalTo = (EqualTo) input;
+            Set<Slot> leftInputSlots = equalTo.left().getInputSlots();
+            Set<Slot> rightInputSlots = equalTo.right().getInputSlots();
+            // not support a=1 b=1 -> a=b, because Sometimes there are cases where the predicates a=1 and a=2
+            // are not eliminated in time. After being pulled up, it is wrong to deduce a=b with b=1.
+            if (leftInputSlots.isEmpty() || rightInputSlots.isEmpty()) {
+                continue;
+            }
+            PredicateInferUtils.getPairFromCast((ComparisonPredicate) input)
+                    .filter(pair -> pair.first instanceof Slot && pair.second instanceof Slot)
+                    .ifPresent(pair -> {
+                        Slot left = (Slot) pair.first;
+                        Slot right = (Slot) pair.second;
+                        fromCastEqualSetBuilder.addEqualPair(left, right);
+                        equalPairs.add(left.getExprId().asInt() <= right.getExprId().asInt()
+                                ? Pair.of(left, right) : Pair.of(right, left));
+                    });
+        }
+        return fromCastEqualSetBuilder.build();
+    }
+
+    /** This is the exposed interface. Inputs are the input predicates for derivation.
```

Comment:
could this equal version infer function also use the graph based algr?  No need to write this specific getEqualSetAndDoReplace to the infer action?

Reply:
same as the deduceTransitiveEquality